man/notify-selfcontained-example.c

```diff
@@ -0,0 +1,167 @@
+/* SPDX-License-Identifier: MIT-0 */
+
+/* Implement the systemd notify protocol without external dependencies.
+ * Supports both readiness notification on startup and on reloading,
+ * according to the protocol defined at:
+ * https://www.freedesktop.org/software/systemd/man/latest/sd_notify.html
+ * This protocol is guaranteed to be stable as per:
+ * https://systemd.io/PORTABILITY_AND_STABILITY/ */
+
+#include <assert.h>
+#include <errno.h>
+#include <inttypes.h>
+#include <signal.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <time.h>
+#include <unistd.h>
+
+#define _cleanup_(f) __attribute__((cleanup(f)))
+
+bool reloading = 0;
+bool terminating = 0;
+
+static void closep(int *fd) {
+  if (!fd || *fd < 0)
+    return;
+
+  close(*fd);
+  *fd = -1;
+}
+
+static int notify(const char *message) {
+  size_t path_length, message_length;
+  struct sockaddr_un socket_addr = {
+    .sun_family = AF_UNIX,
+  };
+  _cleanup_(closep) int fd = -1;
+  const char *socket_path;
+
+  assert(message);
+
+  message_length = strlen(message);
+  if (message_length == 0) {
+    perror("Message to send to $NOTIFY_SOCKET is empty");
+    return -EINVAL;
+  }
+
+  socket_path = getenv("NOTIFY_SOCKET");
+  if (!socket_path)
+    return 0; /* Not running under systemd? Nothing to do */
+
+  path_length = strlen(socket_path);
+  /* Ensure there is room for NULL byte */
```

Comment:
NULL → NUL

(that's the abbreviation ASCII 1967 calls the 0 character. Also see comment here: https://en.wikipedia.org/wiki/Null_character "It is often abbreviated as NUL (or NULL, though in some contexts that term is used for the null pointer.)